### PR TITLE
Azure upgrade check

### PR DESCRIPTION
**Pull request by whites11**
Towards: https://github.com/giantswarm/giantswarm/issues/11102

This PR adds 2 new validation webhooks.
Those webhooks are used to check the upgrade path of Azure Legacy (pre-node pools) Clusters ensuring basic rules are followed.

Rules are:

- can't upgrade the cluster if it's in the creation state
- can't upgrade the cluster if it's in the updating state
- can't skip minor or major releases (but allow skipping patch releases)

### CODE REVIEW

File: pkg/validator/handler.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+			return
+		}
+
+		review := v1beta1.AdmissionReview{}
+		if _, _, err := Deserializer.Decode(data, nil, &review); err != nil {
+			validator.Log("level", "error", "message", "unable to parse admission review request")
+			writer.WriteHeader(http.StatusBadRequest)
+			return
+		}
+		//resourceName := fmt.Sprintf("%s %s/%s", review.Request.Kind, review.Request.Namespace, extractName(review.Request))
```
Comment (njuettner) [inline pkg/validator/handler.go:49]:
```suggestion

```

File: pkg/azureupdate/admit_azureclusterconfig.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
+package azureupdate
```
Comment (njuettner) [inline pkg/azureupdate/admit_azureclusterconfig.go:1]: shouldn't this file called validate_azureclusterconfig.go?

Comment (njuettner) [reply pkg/azureupdate/admit_azureclusterconfig.go:1]: same as the other files?

File: pkg/azureupdate/admit_azureclusterconfig.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+			RestConfig: restConfig,
+		}
+
+		k8sClient, err = k8sclient.NewClients(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	admitter := &AzureClusterConfigValidator{
```
[the remark this answers is not in the document]
Comment (whites11) [reply pkg/azureupdate/admit_azureclusterconfig.go:54]: yeah makes sense, done

File: pkg/azureupdate/admit_azureclusterconfig.go (hunk 2 above)
Comment (whites11) [reply pkg/azureupdate/admit_azureclusterconfig.go]: makes sense as well, done

**Comment by xh3b4sd** [comment]
I did not follow the design decisions we were talking about recently. Is it that we now want to put all provider relevant checks into one project?

**Comment by whites11** [comment]
> I did not follow the design decisions we were talking about recently. Is it that we now want to put all provider relevant checks into one project?

The decision is not made yet.
There is a card in sig releng's board and we will start the discussion during next sync.
Since this feature is blocking azure 12.0.0 release, we (celestial) decided to use what's existing and take the responsibility of refactoring it after the decision will be made.

Are you satisfied about this decision or is this a hard blocker for you @xh3b4sd ?

**Comment by xh3b4sd** [comment]
I do not mind either decision. What is important to me is that we make a conscious decision so that we kind of have an idea about what we are doing. To me it sounds like this is the case. So I am fine. ❤️

File: pkg/azureupdate/common.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+		if r.EQ(*release) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func upgradeAllowed(g8sclient versioned.Interface, oldVersion *semver.Version, newVersion *semver.Version) (bool, error) {
+	if !oldVersion.Equals(*newVersion) {
```
Comment (fiunchinho) [inline pkg/azureupdate/common.go:53]: this function allows _upgrading_ to the same version. Is that what we want?

File: pkg/azureupdate/common.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+		}
+		ret = append(ret, &parsed)
+	}
+
+	return ret, nil
+}
+
+func clusterIsUpgrading(cr *v1alpha1.AzureConfig) (bool, string) {
+	for _, cond := range cr.Status.Cluster.Conditions {
+		if cond.Type == conditionUpdating {
```
Comment (fiunchinho) [inline pkg/azureupdate/common.go:31]:
```
if cond.Type == conditionUpdating {
	return true, conditionUpdating
}
if cond.Type == conditionCreating {
        return true, conditionCreating
}
```

maybe this is simpler?

```
if cond.Type == conditionCreating || cond.Type == conditionUpdating {
    return true, cond.Type
}
```

File: pkg/azureupdate/common.go (hunk 4 above)
Comment (whites11) [reply pkg/azureupdate/common.go:53]: If you don't allow that you are not able to edit the Cars ever again after you create them

**Pull request merged**